closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -0,0 +1,1053 @@
+            String type = service.getType();
+            String algorithm = service.getAlgorithm();
+
+            for (Constraint constraint : constraints) {
+
+                String cType = constraint.type;
+                String cAlgorithm = constraint.algorithm;
+                String cAttribute = constraint.attributes;
+
+                boolean cTypePut = isAsterisk(cType)
+                        || type.equals(cType);
+                boolean cAlgorithmPut = isAsterisk(cAlgorithm)
+                        || algorithm.equals(cAlgorithm);
+                boolean cAttributePut = isAsterisk(cAttribute);
+
+                // For type and algorithm match, and attribute is *.
+                if (cTypePut && cAlgorithmPut && cAttributePut) {
+                    if (debug != null) {
+                        debug.println("Security constraints check."
+                                + " Service type: " + type
+                                + " Algorithm " + algorithm
+                                + " is allowed in provider " + providerName);
+                    }
+                    return true;
+                }
+
+                // For type and algorithm match, and attribute is not *.
+                // Then continue checking attributes.
+                if (cTypePut && cAlgorithmPut) {
+                    String[] cAttributeArray = constraint.attributes.split(":");
+
+                    // For each attribute, must be all matched for return allowed.
+                    for (String attribute : cAttributeArray) {
+                        String[] input = attribute.split("=", 2);
+
+                        String cName = input[0].trim();
+                        String cValue = input[1].trim();
+                        String sValue = service.getAttribute(cName);
+                        if ((sValue == null) && !cValue.equalsIgnoreCase(sValue)) {
+                            // Any of the attribute not match, return service is not allowed
+                            return false;
+                        }
+                    }
+                    if (debug != null) {
+                        debug.println(
+                                "Security constraints check."
+                                        + " Service type: " + type
+                                        + " Algorithm: " + algorithm
+                                        + " Attribute: " + constraint.attributes
+                                        + " is allowed in provider: " + providerName);
+                    }
+                    return true;
+                }
+            }
+            if (debug != null) {
+                debug.println("Security constraints check."
+                        + " Service type: " + type
+                        + " Algorithm: " + algorithm
+                        + " is NOT allowed in provider " + providerName);
+            }
+            // Go through all the constraints for the provider.
```

Comment:
This comment doesn't read well; perhaps it should simply be deleted (I think line 816 is sufficient).